cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredToAvroTransformer.java

```diff
@@ -47,7 +51,112 @@ public GenericRecord transform(StructuredRecord structuredRecord) throws Excepti
     GenericRecordBuilder recordBuilder = new GenericRecordBuilder(avroSchema);
     for (Schema.Field field : avroSchema.getFields()) {
       String fieldName = field.name();
-      recordBuilder.set(fieldName, structuredRecord.get(fieldName));
+      recordBuilder.set(fieldName, convertField(getRecordField(structuredRecord, fieldName), field.schema()));
+    }
+    return recordBuilder.build();
+  }
+
+  private Object convertField(Object field, Schema fieldSchema) {
+    Schema.Type fieldType = fieldSchema.getType();
+    switch (fieldType) {
+      case RECORD:
+        return convertRecord(field, fieldSchema);
+      case ARRAY:
+        return convertArray(field, fieldSchema.getElementType());
+      case MAP:
+        return convertMap((Map<String, Object>) field, fieldSchema.getValueType());
+      case UNION:
+        return convertUnion(field, fieldSchema.getTypes());
+      case NULL:
+        return null;
+      case STRING:
```

Comment:
for string, we should return field.toString() since avro doesn't always use a String to represent a String type